allennlp/modules/bimpm_matching.py

```diff
@@ -0,0 +1,396 @@
+"""
+Multi-perspective matching layer
+"""
+
+from typing import Tuple, List
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from allennlp.common.checks import ConfigurationError
+from allennlp.common.registrable import FromParams
+from allennlp.nn.util import get_lengths_from_binary_sequence_mask
+
+
+def masked_max(vector: torch.Tensor,
+               mask: torch.Tensor,
+               dim: int,
+               keepdim: bool = False,
+               min_val: float = -1e7) -> torch.Tensor:
+    """
+    To calculate max along certain dimensions on masked values
+
+    Parameters
+    ----------
+    vector : ``torch.Tensor``
+        The vector to calculate max, assume unmasked parts are already zeros
+    mask : ``torch.Tensor``
+        The mask of the vector. It must be broadcastable with vector.
+    dim : ``int``
+        the dimension to calculate max
+    keepdim : ``bool``
+        Whether to keep dimension
+    min_val : ``float``
+        The minimal value for paddings
+
+    Returns
+    -------
+    A ``torch.Tensor`` of including the maximum values.
+    """
+    replaced_vector = vector + (1.0 - mask) * min_val
+    max_value, _ = replaced_vector.max(dim=dim, keepdim=keepdim)
+    return max_value
+
+
+def masked_mean(vector: torch.Tensor,
+                mask: torch.Tensor,
+                dim: int,
+                keepdim: bool = False,
+                eps: float = 1e-8) -> torch.Tensor:
+    """
+    To calculate mean along certain dimensions on masked values
+
+    Parameters
+    ----------
+    vector : ``torch.Tensor``
+        The vector to calculate mean, assume unmasked parts are already zeros
```

Comment:
For a general utility, I wouldn't make this assumption, instead doing a `vector.masked_fill_(0, mask)` before the `torch.sum()`.